README.md
[![Build status — Travis-CI][travis-icon]][travis]
[![cookie-notice on Npmjs][npm-icon]][npm]
[![License][license-icon]][mit]
[![Total downloads ~ Npmjs][downl-icon]][npm]
[![Size of Javascript][size-icon]][build]

# CookieNotice

**CookieNoticeJS** is a very simple and small *(→ 2 kB gzip)* vanilla JS script with multi language support for [GDPR][]/[DSGVO][]‎ transparency and
notification purposes that provides an easy way to show a cookie notice on your website.

<img src="https://i.imgur.com/koDf1h0.png" alt="FetchBot" align="center"/>

**Available via [npm][]**

```shell
npm install cookie-notice
npm test
```


**To use in your project There are plenty ways for integration:**

**When installed via npm, include in any project by using path below:**
```html
<script src="node-modules/cookie-notice/dist/cookie.notice.min.js"></script>
```

**For usage undeled with Angular2+ add line below in "scripts" section in angular-cli.json:**
```json
{
    "scripts": [
        "../node_modules/cookie-notice/dist/cookie.notice.js",
        "../optional/path/to/custom/cookie-notice-config.js"
    ]
}
```

**When cloned directly from gitHub use path below:**
```html
<script src="cookie-notice/dist/cookie.notice.min.js"></script>
```

## Behavior
You will get a dismissable banner on the bottom of your pages showing a default cookie audit like the following:

> We use cookies to make sure you can have the best experience on our website. If you continue to use this site we assume that you will be happy with it.

Check my website for a [DEMO](http://codeb.it/). Depending on the visitor browser language one of the preloaded translations will be shown. At the moment **CookieNoticeJS** supports *EN, IT, DE and FR*. If you want to contribute with an extra language do not hesitate to open an issue or a PR.

**CookieNoticeJS** has been successfully tested on IE9+, Chrome, Firefox and Safari.

## Customize CookieNoticeJS

For the most of you including the script should be enough but **CookieNoticeJS** comes with many customization options. Let's see an example:

```html
<script src="js/cookie.notice.min.js"></script>
<script>
    new cookieNoticeJS({

       // Localizations of the notice message
       'messageLocales': {
         'it': 'Custom localized message'
       },

       // Localizations of the dismiss button text
       'buttonLocales': {
         'it': 'Chiudi'
       },

       // Position for the cookie-notifier (default=bottom)
       'cookieNoticePosition': 'top',

       // Shows the "learn more button (default=false)
       'learnMoreLinkEnabled': false,

       // The href of the learn more link must be applied if (learnMoreLinkEnabled=true)
       'learnMoreLinkHref': '/learn/more/index.html',

       // Text for optional learn more button
       'learnMoreLinkText':{
           'en':'learn more'
       },

       // The message will be shown again in X days
       'expiresIn': 30,

       // DEPRECATED.
       'buttonBgColor': '#d35400',
       'buttonTextColor': '#fff',
       'noticeBgColor': '#000',
       'noticeTextColor': '#fff',
       'linkColor': '#f00',
       // END DEPRECATED.

       'styleLinkHref': '../cookie-notice.css',

       // The target of the learn more link (default='', or '_blank')
       'linkTarget': '',

       // Print debug output to the console (default=false)
       'debug': false
    });
</script>
```

### Deprecated properties

In `cookie-notice` >= Version 2.x the color options shown above will be removed
(`buttonBgColor`, `buttonTextColor`, `noticeBgColor`, `noticeTextColor`, `linkColor`).
Instead, developers and designers can override these properties using the CSS styles for the page.

## Configuration via `data-` attribute

Configuration options can be put in a `data-cookie-notice` HTML attribute in JSON format.
Note, you can include the Javascript from the [unpkg][] CDN ([browse][]).
For example:

```html
<script
  data-cookie-notice='{ "learnMoreLinkEnabled": true, "learnMoreLinkHref": "/privacy.html" }'
  src="https://unpkg.com/cookie-notice@^1/dist/cookie.notice.min.js"
></script>
```

#### Author
Alessandro Benoit

#### Contributors
[Bernhard Behrendt](mailto:bernhard.behrendt@aoe.com) [@AOEpeople](https://github.com/AOEpeople),
[Nick Freear](https://github.com/nfreear) [IET at the OU](https://github.com/IET-OU)

### License

License: [MIT][]


[MIT]: https://github.com/micc83/cookie-notice-js/blob/master/LICENSE
    "License: MIT | Copyright © 2018 Alessandro Benoit (micc83)."
[MIT-0]: https://mit-license.org/#2018
[travis-icon]: https://travis-ci.org/AOEpeople/cookie-notice.svg?branch=master
[travis]: https://travis-ci.org/AOEpeople/cookie-notice "Build status – Travis-CI"
[npm]: https://npmjs.com/package/cookie-notice "CookieNotice – on NPM"
[npm-icon]: https://badge.fury.io/js/cookie-notice.svg
[npm-i0]: https://img.shields.io/npm/v/cookie-notice.svg "(Timeout errors)"
[license-icon]: https://img.shields.io/npm/l/cookie-notice.svg
[downl-icon]: https://img.shields.io/npm/dt/cookie-notice.svg "Count of total downloads – NPM"
[build]: https://github.com/AOEpeople/cookie-notice/tree/master/dist
[size-icon]: https://img.shields.io/github/size/AOEpeople/cookie-notice/dist/cookie.notice.min.js.svg
    "Size of built Javascript, kilo-bytes (kB) – on GitHub"
[unpkg]: https://unpkg.com/ "unpkg is a fast, global content delivery network (CDN) for everything on npm."
[browse]: https://unpkg.com/cookie-notice@^1/ "Browse cookie-notice on unpkg"

[DSGVO]: https://de.wikipedia.org/wiki/Datenschutz-Grundverordnung "Datenschutz-Grundverordnung (DSGVO)"
[GDPR]: https://en.wikipedia.org/wiki/General_Data_Protection_Regulation "General Data Protection Regulation (GDPR)"

[End]: //.


CHANGELOG.md

# CookieNotice changelog

← [README][]

Version History

## Version 2.0.0

 * _Date:  approx. June 2018_;
 * Move CSS from `style` attributes to `<style>` or `<link rel=stylesheet ... >`, Issue [#8][];
 * The following configuration options are __removed__ in this version:
   `buttonBgColor`, `buttonTextColor`, `noticeBgColor`, `noticeTextColor`, `linkColor`.

## Version 1.1.12

 * _Date:  approx. 4-30 June 2018_;
 * Add accessibility testing via [grunt-rsids-pa11y][] and [pa11y][], WCAG 2.0 AA, pull #6;
 * Accessibility fixes — colour contrast, button role, link target, link underline;
 * Accessibility fixes — express CSS font-size in `em` or `rem`, not `px`;
 * Add configuration option — `linkTarget` (default: '');
 * Edit the language used in the English translation;
 * Fix `browser` and `repository` fields in `package.json`;
 * Add `files` and `bugs` fields to `package.json`;
 * Add `engines` and `contributors` fields to `package.json`;
 * Add `live-server` to `peerDependencies` in `package.json`;
 * Remove Nick's email address from `README` & `bower.json` (!)

## Version 1.1.11

 * _Date:  30 May 2018_;
 * Add support for `data-cookie-notice` attribute on `<script>`, pull #5, Issue [#4][];
 * Document `data-` attribute in README;
 * Document _unpkg_ CDN in README;

## Version 1.1.10

 * _Date:  29 May 2018_;
 * Move jQuery to `devDependencies` for clarity, pull #2;
 * Point `tests/index.html` to local copies of jQuery and QUnit (was CDN);

## Version 1.1.9

 * _Date:  25 May 2018_;
 * Add Travis-CI testing to the project, pull #1;
 * Update test dependencies —
   jQuery from 2.x → 3.x; QUnit from 1.x → 2.x;
 * Add SVG badges for Travis-CI, NPMJS.com, etc. to `README.md`;

## Version 1.1.8

 * _Date:  25 May 2018_;

## Version 1.1.7

 * _Date:  25 April 2018_;

...

## Version 1.1.0

 * _Date:  24 April 2018_;
 * First release on [npm][];

## Version 1.0.0

 * _Date:  20 May 2015_;
 * Release by:  [@micc83][];
 * (Not on [npm][])

---

← [README][]

[readme]: https://github.com/AOEpeople/cookie-notice#readme
[npm]: https://www.npmjs.com/package/cookie-notice
[@micc83]: https://github.com/micc83/cookie-notice-js "Alessandro Benoit"
[#4]: https://github.com/AOEpeople/cookie-notice/issues/4
    "'data-cookie-notice' HTML attribute, other suggestions / wishlist"
[#8]: https://github.com/AOEpeople/cookie-notice/issues/8
    "Move styles to separate stylesheet, included via Javascript"
[a11y-fail]: https://travis-ci.org/nfreear/cookie-notice/jobs/387344672#L1217-L1235
[grunt-rsids-pa11y]: https://www.npmjs.com/package/grunt-rsids-pa11y
[pa11y]: http://pa11y.org/


tests/tests.js
/**
 * Clear cookies
 */
function clearCookies() {
    document.cookie.split(";").forEach(function (c) {
        document.cookie = c.replace(/^ +/, "").replace(/=.*/, "=;expires=" + new Date().toUTCString() + ";path=/");
    });
}

// http://qunitjs.com/cookbook/#custom-assertions
QUnit.assert.contains = function( haystack, needle, message ) {
  var actual = haystack.indexOf(needle) > -1;
  this.pushResult({ result: actual, actual: actual, expected: needle, message: message });
  // Was: this.push(actual, actual, needle, message);
};

QUnit.assert.matches = function( haystack, regExp, message ) {
  var result = regExp.test( haystack );
  this.pushResult({ result: result, actual: result, expected: regExp, message: message });
};

if (window._phantom) {
    if (!HTMLElement.prototype.click) {
        HTMLElement.prototype.click = function () {
            var ev = document.createEvent('MouseEvent');
            ev.initMouseEvent(
                'click',
                /*bubble*/true, /*cancelable*/true,
                window, null,
                0, 0, 0, 0, /*coordinates*/
                false, false, false, false, /*modifier keys*/
                0/*button=left*/, null
            );
            this.dispatchEvent(ev);
        };
    }
}

clearCookies();

QUnit.test('Check if extendDefaults works correctly', function (assert) {

    var result = cookieNoticeJS.extendDefaults({
            'foo': 'bar',
            'fooo': {
                'f1': 'Hello',
                'f2': 'Ciao Mondo'
            }
        },
        {
            'foo': 'barFoo',
            'fooo': {
                'f1': 'Hello World'
            }
        });

    assert.deepEqual(result, {
        "foo": "barFoo",
        'fooo': {'f1': 'Hello World', 'f2': 'Ciao Mondo'}
    }, 'extendDefaults working properly');

});

$(document).ready(function () {

    QUnit.test('Check if the notice is present', function (assert) {
        assert.strictEqual($('#cookieNotice').length, 1, 'Notice is present');
        assert.strictEqual($('#cookieNotice').find('a.confirm').length, 1, 'Notice button is present');
        assert.contains($('#cookieNotice')[0].className, 'cookie-notice-js', 'the className contains "cookie-notice-js"');
    });

    QUnit.test('Check that notice disappears when clicking on the OK button', function (assert) {

        assert.expect(2);

        var done = assert.async();

        $('#cookieNotice').find('a.confirm')[0].click();

        window.setTimeout(function () {
            assert.strictEqual($('#cookieNotice').length, 0, 'Notice is gone');
            assert.notEqual(document.cookie.indexOf('cookie_notice'), -1, 'Cookie has been set');
            done();
        }, 500);

    });

    QUnit.test('Check that notice doesn\'t show up when cookie is set', function (assert) {

        cookieNoticeJS.clearInstance();
        new cookieNoticeJS();
        assert.strictEqual($('#cookieNotice').length, 0, 'Notice should not be present');

    });

    QUnit.test('And that it shows up when the cookie are cleared', function (assert) {

        clearCookies();
        cookieNoticeJS.clearInstance();
        new cookieNoticeJS();
        assert.strictEqual($('#cookieNotice').length, 1, 'Notice should be present');
    });

    QUnit.test('And there is no "learn more" link in default configuration case', function (assert) {

        clearCookies();
        cookieNoticeJS.clearInstance();
        new cookieNoticeJS();
        assert.strictEqual($('a.learn-more').length, 0, 'Learn more link shouldn\'t be present');
    });

    QUnit.test('And cookie-notice is stuck to the bottom', function (assert) {
        clearCookies();
        cookieNoticeJS.clearInstance();
        new cookieNoticeJS();

        assert.matches($('#cookieNotice').attr('class'), / bottom/, 'the className contains "bottom"' );

        assert.strictEqual($('#cookieNotice').css('bottom'), '0px', 'positioning is stuck to the bottom');
        assert.strictEqual($('#cookieNotice').css('top'), 'auto', 'so the top position is "auto"');
    });

    QUnit.test('When defaults have overrides changes should take place', function (assert) {

        $('#cookieNotice').remove();

        var mockedLink = 'some/fancy/link.html';

        clearCookies();
        cookieNoticeJS.clearInstance();
        new cookieNoticeJS({
            learnMoreLinkEnabled: true,
            learnMoreLinkHref: mockedLink
        });

        var learnMoreLink = $('a.learn-more');

        assert.strictEqual(learnMoreLink.length, 1, 'Learn more link should be present');
        assert.strictEqual(learnMoreLink.attr('href'), mockedLink, 'Learn more link href equals mocked link');
        assert.strictEqual(learnMoreLink.attr('target'), '', 'The target of the learn-more link should be empty (default)');
    });
});


gruntfile.js
module.exports = function (grunt) {

    // Project configuration.
    grunt.initConfig({
        pkg: grunt.file.readJSON('package.json'),
        connect: {
            server: {
                options: {
                    port: 8000,
                    base: '.'
                }
            }
        },
        less: {
            dist: {
                options: {
                    banner:
                    '/*! <%= pkg.name %> v<%= pkg.version %> by <%= pkg.author %> <%= grunt.template.today("yyyy-mm-dd") %>.\n\n*/\n',
                    compress: true
                    // paths: [ 'assets/css' ]
                },
                files: {
                    'dist/cookie.notice.css': 'style/style.less'
                }
            }
        },
        qunit: {
            all: {
                options: {
                    urls: [
                        'http://localhost:8000/tests/index.html'
                    ]
                }
            }
        },
        rsids_pa11y: { // Accessibility testing, via ~ http://pa11y.org/
            test: {
                options: { // Task-specific options go here.
                    // screenCapture: './_pa11y-screen-capture.png',
                    standard: 'WCAG2AA', // Or 'WCAG2AAA'
                    timeout: 5000,
                    wait: 500,
                    rootElement: 'body', // Was: '#cookieNotice',
                    verifyPage: 'id="cookieNotice"' // Not supported?
                },
                url: [ 'http://localhost:8000/tests/data.html' ]
                //, file: [ 'array of files, globbing permitted' ]
            }
        },
        strip_code: {

            src: {
                src: 'src/cookie.notice.js',
                dest: 'dist/cookie.notice.js'
            }
        },
        uglify: {
            options: {
                banner: '/*! <%= pkg.name %> v<%= pkg.version %> by <%= pkg.author %> <%= grunt.template.today("yyyy-mm-dd") %> */\n',
                report: 'gzip',
                screwIE8: true
            },
            build: {
                src: 'dist/cookie.notice.js',
                dest: 'dist/cookie.notice.min.js'
            }
        }
    });

    // Load the plugins
    grunt.loadNpmTasks('grunt-contrib-less');
    grunt.loadNpmTasks('grunt-contrib-uglify');
    grunt.loadNpmTasks('grunt-contrib-connect');
    grunt.loadNpmTasks('grunt-contrib-qunit');
    grunt.loadNpmTasks('grunt-strip-code');
    grunt.loadNpmTasks('grunt-rsids-pa11y');

    // Default task(s).
    grunt.registerTask('default', [ 'test', 'strip_code', 'uglify', 'less' ]);
    grunt.registerTask('test', [ 'connect', 'qunit', 'rsids_pa11y' ]);

};


src/cookie.notice.js
/**
 * Cookie Notice JS
 * @author Alessandro Benoit
 */
(function () {

    "use strict";

    /**
     * Store current instance
     */
    var instance;
    // originPaddingTop;

    /**
     * Defaults values
     * @type object
     */
    var defaults = {
        'messageLocales': {
            'it': 'Utilizziamo i cookie per essere sicuri che tu possa avere la migliore esperienza sul nostro sito. Se continui ad utilizzare questo sito assumiamo che tu ne sia felice.',
            'en': 'We use cookies to ensure that you have the best experience on our website. If you continue to use this site we assume that you accept this.',
            'de': 'Wir verwenden Cookies um sicherzustellen dass Sie das beste Erlebnis auf unserer Website haben.',
            'fr': 'Nous utilisons des cookies afin d\'être sûr que vous pouvez avoir la meilleure expérience sur notre site. Si vous continuez à utiliser ce site, nous supposons que vous acceptez.'
        },

        'cookieNoticePosition': 'bottom',

        'learnMoreLinkEnabled': false,

        'learnMoreLinkHref': '/cookie-banner-information.html',

        'learnMoreLinkText': {
            'it': 'Saperne di più',
            'en': 'Learn more',
            'de': 'Mehr erfahren',
            'fr': 'En savoir plus'
        },

        'buttonLocales': {
            'en': 'OK'
        },

        'expiresIn': 30,

        /* DEPRECATED. No longer needed!
        'buttonBgColor': '#ca5000', // Accessibility contrast fix (Was: '#d35400') (WCAG2AAA: '#983c00').
        'buttonTextColor': '#fff',
        'noticeBgColor': '#000',
        'noticeTextColor': '#fff',
        'linkColor': '#009fdd',
        END DEPRECATED. */

        'styleLinkHref': '/../../dist/cookie.notice.css',
        'scriptSelector': 'script[ src *= "cookie.notice." ]',

        'linkTarget': '', // Accessibility fix (Was: '_blank').
        'debug': false
    };

    /**
     * Initialize cookie notice on DOMContentLoaded
     * if not already initialized with alt params
     */
    document.addEventListener('DOMContentLoaded', function () {
        if (!instance) {
            new cookieNoticeJS();
        }
    });

    /**
     * Constructor
     * @constructor
     */
    window.cookieNoticeJS = function () {

        // If an instance is already set stop here
        if (instance !== undefined) {
            return;
        }

        // Set current instance
        instance = this;

        // If cookies are not supported or notice cookie is already set
        if (getNoticeCookie()) {
            return;
        }

        // 'data-' attribute - data-cookie-notice='{ "key": "value", ... }'
        var elemCfg = document.querySelector('script[ data-cookie-notice ]');
        var config;
        try {
            config = elemCfg ? JSON.parse(elemCfg.getAttribute('data-cookie-notice')) : {};
        } catch (ex) {
            console.error('data-cookie-notice JSON error:', elemCfg, ex);
            config = {};
        }

        // Extend default params
        var params = extendDefaults(defaults, arguments[0] || config || {});

        if (params.debug) {
            console.warn('cookie-notice:', params);
        }

        addStylesheetLink(params.styleLinkHref, params.scriptSelector);

        // Get current locale for notice text
        var noticeText = getStringForCurrentLocale(params.messageLocales);

        // Create notice
        var notice = createNotice(noticeText, params.cookieNoticePosition);

        var learnMoreLink;

        if (params.learnMoreLinkEnabled) {
            var learnMoreLinkText = getStringForCurrentLocale(params.learnMoreLinkText);

            learnMoreLink = createLearnMoreLink(learnMoreLinkText, params.learnMoreLinkHref, params.linkTarget);
        }

        // Get current locale for button text
        var buttonText = getStringForCurrentLocale(params.buttonLocales);

        // Create dismiss button
        var dismissButton = createDismissButton(buttonText);

        // Dismiss button click event
        dismissButton.addEventListener('click', function (e) {
            e.preventDefault();
            setDismissNoticeCookie(parseInt(params.expiresIn + "", 10) * 60 * 1000 * 60 * 24);
            fadeElementOut(notice);
        });

        // Append notice to the DOM
        var noticeDomElement = document.body.appendChild(notice);

        if (!!learnMoreLink) {
            noticeDomElement.appendChild(learnMoreLink);
        }

        noticeDomElement.appendChild(dismissButton);
    };

    /**
     * Get the string for the current locale
     * and fallback to "en" if none provided
     * @param locales
     * @returns {*}
     */
    function getStringForCurrentLocale(locales) {
        var locale = (navigator.userLanguage || navigator.language).substr(0, 2);
        return (locales[locale]) ? locales[locale] : locales['en'];
    }

    /**
     * Test if notice cookie is there
     * @returns {boolean}
     */
    function getNoticeCookie() {
        return document.cookie.indexOf('cookie_notice') != -1;
    }

    /**
     * Create notice
     * @param message
     * @param position
     * @returns {HTMLElement}
     */
    function createNotice(message, position) {

        var notice = document.createElement('div');
        /* var lineHeight = 2, // Was: 28 (px).
            paddingBottomTop = 10,
            fontSize = lineHeight / 2.333,
            noticeHeight = lineHeight + paddingBottomTop * 2; */

        notice.innerHTML = message + '&nbsp;';
        notice.setAttribute('id', 'cookieNotice');
        notice.setAttribute('data-test-section', 'cookie-notice');
        notice.setAttribute('data-test-transitioning', 'false');

        notice.className = 'cookie-notice-js ';

        notice.className += (position === 'top') ? 'top' : 'bottom';

        if (position === 'top') {
            var bodyDOMElement = document.querySelector('body');

            bodyDOMElement.className += ' cookie-notice-js-body-top';

            // originPaddingTop = bodyDOMElement.style.paddingTop;
        }

        return notice;
    }

    /**
     * Create dismiss button
     * @param message
     * @returns {HTMLElement}
     */
    function createDismissButton(message) {

        var dismissButton = document.createElement('a');

        // Dismiss button
        dismissButton.href = '#';
        dismissButton.innerHTML = message;

        dismissButton.setAttribute('role', 'button'); // Accessibility fix.
        dismissButton.className = 'confirm';

        dismissButton.setAttribute('data-test-action', 'dismiss-cookie-notice');

        return dismissButton;
    }

    /**
     * Create the 'learn more' link
     *
     * @param learnMoreLinkText
     * @param learnMoreLinkHref
     * @param linkTarget
     * @returns {HTMLElement}
     */
    function createLearnMoreLink(learnMoreLinkText, learnMoreLinkHref, linkTarget) {

        var learnMoreLink = document.createElement('a');

        learnMoreLink.href = learnMoreLinkHref;
        learnMoreLink.textContent = learnMoreLinkText;
        learnMoreLink.target = linkTarget;
        learnMoreLink.className = 'learn-more';
        learnMoreLink.setAttribute('data-test-action', 'learn-more-link');

        return learnMoreLink;
    }

    /**
     * Set sismiss notice cookie
     * @param expireIn
     */
    function setDismissNoticeCookie(expireIn) {
        var now = new Date(),
            cookieExpire = new Date();

        cookieExpire.setTime(now.getTime() + expireIn);
        document.cookie = "cookie_notice=1; expires=" + cookieExpire.toUTCString() + "; path=/;";
    }

    /**
     * Fade a given element out
     * @param element
     */
    function fadeElementOut(element) {
        element.style.opacity = 1;

        element.setAttribute('data-test-transitioning', 'true');

        (function fade() {
            if ((element.style.opacity -= .1) < 0.01) {

                /* if (originPaddingTop !== undefined) {
                    var bodyDOMElement = document.querySelector('body');
                    bodyDOMElement.style.paddingTop = originPaddingTop;
                } */

                document.body.removeChild(element);
            } else {
                setTimeout(fade, 40);
            }
        })();
    }

    /**
     * Utility method to extend defaults with user options
     * @param source
     * @param properties
     * @returns {*}
     */
    function extendDefaults(source, properties) {
        var property;
        for (property in properties) {
            if (properties.hasOwnProperty(property)) {
                if (typeof source[property] === 'object') {
                    source[property] = extendDefaults(source[property], properties[property]);
                } else {
                    source[property] = properties[property];
                }
            }
        }
        return source;
    }

    /**
     * Method to add a stylesheet <link> to the page.
     * @param styleUrl
     * @param scriptSelector
     */
    function addStylesheetLink(styleUrl, scriptSelector) {
        if (!styleUrl) return;

        // Assume the Javascript and stylesheet are on a different domain to the page (e.g. CDN).
        var styleLink = document.createElement('link');
        var script = document.querySelector(scriptSelector);
        var href = (script ? script.src : '') + styleUrl;

        styleLink.rel = 'stylesheet';
        styleLink.type = 'text/css';
        styleLink.href = href;

        document.head.appendChild(styleLink);
    }

    /* test-code */
    cookieNoticeJS.extendDefaults = extendDefaults;
    cookieNoticeJS.clearInstance = function () {
        instance = undefined;
    };
    /* end-test-code */

}());
